internal/perms/acl.go

```diff
@@ -209,6 +220,69 @@ func (a ACL) Allowed(r Resource, aType action.Type, userId string, opt ...Option
 	return
 }
 
+// AllowedListConditions builds a set of user permissions (according to their grants)
+// for resource listing ahead of DB access.
+// This enables list query optimization: instead of having to query all data and then drop
+// what the requestor is not authorized to see, we can use the data in the Permission objects
+// to build a more efficient DB query.
+// It does not (and is not meant to) override current authorization mechanisms.
+func (a ACL) AllowedListConditions(scopes map[string]*scopes.ScopeInfo, r resource.Type, aSet action.ActionSet) []ListPermission {
```

Comment:
Then this could be:

```suggestion
// ListPermissions builds a set of Permissions based on the grants in the ACL for the List action for the given scpoes and resource type. 
func (a ACL) ListPermissions(scopes map[string]*scopes.ScopeInfo, r resource.Type, aSet action.ActionSet) []ListPermission {
```

Reply:
Changed in [bb61302](https://github.com/hashicorp/boundary/pull/2335/commits/bb61302c23afbdd45259a44f39711299a2fc450e)